File: api-server/src/utils/response.ts
import type { Response } from 'express';
import type { ApiResponse } from '../types/index.js';

export class ApiResponseUtil {
    static success<T>(res: Response, data: T, message?: string, statusCode = 200): Response {
        const response: ApiResponse<T> = {
            success: true,
            data,
            ...(message && { message })
        };
        return res.status(statusCode).json(response);
    }

    static created<T>(res: Response, data: T, message = 'Created successfully'): Response {
        return this.success(res, data, message, 201);
    }

    static error(
        res: Response,
        message: string,
        statusCode = 500,
        error?: string
    ): Response {
        const response: ApiResponse = {
            success: false,
            message,
            ...(error && { error })
        };
        return res.status(statusCode).json(response);
    }

    static badRequest(res: Response, message: string, error?: string): Response {
        return this.error(res, message, 400, error);
    }

    static unauthorized(res: Response, message = 'Unauthorized', error?: string): Response {
        return this.error(res, message, 401, error);
    }

    static forbidden(res: Response, message = 'Access denied', error?: string): Response {
        return this.error(res, message, 403, error);
    }

    static notFound(res: Response, message = 'Resource not found', error?: string): Response {
        return this.error(res, message, 404, error);
    }

    static validationError(res: Response, errors: Array<{ field: string; message: string }>): Response {
        const response: ApiResponse = {
            success: false,
            message: 'Validation failed',
            errors
        };
        return res.status(400).json(response);
    }

    static internalError(res: Response, error?: Error): Response {
        const message = 'Internal Server Error';
        const errorMessage = process.env.NODE_ENV === 'development' ? error?.message : undefined;
        return this.error(res, message, 500, errorMessage);
    }

    static tooManyRequests(res: Response, message = 'Too many requests, please try again later'): Response {
        return this.error(res, message, 429);
    }
}

File: api-server-go/internal/utils/response.go
package utils

import (
	"encoding/json"
	"net/http"
)

type APIResponse struct {
	Success bool        `json:"success"`
	Message string      `json:"message,omitempty"`
	Data    interface{} `json:"data,omitempty"`
	Error   string      `json:"error,omitempty"`
}

func RespondJSON(w http.ResponseWriter, statusCode int, response APIResponse) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(response)
}

func Success(w http.ResponseWriter, data interface{}, message ...string) {
	msg := ""
	if len(message) > 0 {
		msg = message[0]
	}

	RespondJSON(w, http.StatusOK, APIResponse{
		Success: true,
		Message: msg,
		Data:    data,
	})
}

func Created(w http.ResponseWriter, data interface{}, message string) {
	RespondJSON(w, http.StatusCreated, APIResponse{
		Success: true,
		Message: message,
		Data:    data,
	})
}

func Error(w http.ResponseWriter, statusCode int, message string, errorDetail string) {
	RespondJSON(w, statusCode, APIResponse{
		Success: false,
		Message: message,
		Error:   errorDetail,
	})
}

func NotFound(w http.ResponseWriter, message string) {
	Error(w, http.StatusNotFound, message, "Resource not found")
}

func Unauthorized(w http.ResponseWriter, message string) {
	Error(w, http.StatusUnauthorized, message, "Unauthorized access")
}

func Forbidden(w http.ResponseWriter, message string) {
	Error(w, http.StatusForbidden, message, "Access forbidden")
}

func BadRequest(w http.ResponseWriter, message string) {
	Error(w, http.StatusBadRequest, message, "Invalid request")
}

func InternalServerError(w http.ResponseWriter, message string) {
	Error(w, http.StatusInternalServerError, message, "Internal server error")
}
